docs/proposals/operator-sdk-v1-migration.md

```diff
@@ -0,0 +1,285 @@
+2. `operator-sdk build` has likewise been replaced with a make target - `make docker-build` [13].
+3. Deploying is likewise handled by a make targets - `make install` and `make deploy` [14].
+
+The generated Makefile targets can be used by contributors to test Shipwright Builds. We are free
+to add our own targets to generate an "all-in-one" YAML manifest.
+
+[13] https://sdk.operatorframework.io/docs/building-operators/golang/migration/#generate-manifests-and-build-the-operator
+[14] https://sdk.operatorframework.io/docs/building-operators/golang/quickstart/
+
+
+### Risks and Mitigations
+
+**Risk**: Bug fixes and features are wiped out as we migrate to operator-sdk v1.
+
+*Mitigation*: As the rebase PR nears completion, we hold outstanding pull requests until it merges.
+Our CI needs to ensure that the rebase can merge without conflicts.
+
+## Design Details
+
+### Test Plan
+
+Existing test suites would remain in place, with overhauls to utilize envtest for integration testing.
+The generated `make test` target runs `go test ./...`, 
+
+### Graduation Criteria
+
+Not applicable
+
+### Upgrade / Downgrade Strategy
+
+Since this project is in the alpha state, support for in-place upgrade is not a hard requirement.
+This proposal outlines specific areas where there is potential to introduce breaking changes that
+downstream distributions may need to consider.
+
+### Version Skew Strategy
+
+If the leader election ConfigMap is left unchanged, we can support operator skew.
+
+## Implementation History
+
+2020-12-08: Initial proposal
+2020-12-09: Alternative using knative/pkg
+
+## Drawbacks
+
+Migrating to operator-sdk v1 requires significant project restructuring and changes to how we run
+tests. Bug fixes and other enhancements could easily get lost in the shuffle.
+
+The primary benefit of operator-sdk is to facilitate the deployment of Shipwright Build as an
+operator managed by the Operator Lifecycle Manager. operator-sdk has tooling to help create the
+necessary artifacts for OLM, but one does not need operator-sdk to create the components needed to
+deploy an operator via OLM.
+
+## Alternatives
+
+### Use Upstream Components
+
+We can use the sub-components that operator-sdk packages together, without using operator-sdk
+directly:
+
+1. Controller-runtime can be used to write the controllers and controller manager [15].
+2. KubeBuilder can be used to create the custom resources [16].
+3. Kustomize can be used to manage our YAML manifests [17].
+
+[15] https://github.com/kubernetes-sigs/controller-runtime
+[16] https://github.com/kubernetes-sigs/kubebuilder
+[17] https://kustomize.io/
+
+### Migrate code only
+
+In this scenario, we migrate our CRDs and controller code only. Integration and e2e tests remain
+as-is, potentially move to envtest at a later date.
+
+### Switch to knative.dev/pkg
+
+Tekton uses knative packages for their controller code and operator. They are currently shipping
```

Comment:
This one was not very clear to me, what does the "switch" here implies to Build? :)

Reply:
It means instead of using opeator-sdk and dependent libraries (kubeBuilder, controller-runtime), we use `knative/pkg` as the baseline for creating our controllers. This is a refactoring of similar scope and complexity.

Reply:
I´m not 100% sure we need a full switch to knative/pkg. I believe we can rely only in controller-runtime, some pkgs of sdk, but not in the sdk binary. Moving fully to knative/pkg is not that easy, one needs to contemplate:
- knative/pkg is only intended for the knative crds and tekton
- tekton use it, because they were before in knative

I feel relying on knative 100% might have implications in terms of flexibility and code understanding(not everything there is document, but rather more like a tribe specific knowledge), while something like controller-runtime is more generic and well documented.

Reply:
To clarify @adambkaplan 's comment slightly, the *upstream* tekton operator at https://github.com/tektoncd/operator has no operator-sdk elements, but the (original) *downstream* tekton operator at https://github.com/openshift/tektoncd-pipeline-operator still has operator sdk elements ... it is still vendored in (though at v0.17.2 as of my writing).

A thought that occurred to me is that perhaps we could see if @vdemeester and/or @nikhil-thomas could join an upcoming shipwright community meeting (Mondays at 9 AM Eastern US) to provide some details/history/roadmap for the tekton operators they drive and all involved in assessing knative vs. operator-sdk,  migration details, etc.

FWIW, from an "upstream" / "downstream" perspective, we would consider https://github.com/shipwright-io/build "upstream" or "k8s neutral" ... i.e. analogous to https://github.com/tektoncd/operator and https://github.com/tektoncd/pipeline

Reply:
Right, we can definitely join the shipwright community meeting to discuss this 😉 

I tend to agree with @qu1queee, if the tekton project was bootstrap today, without any *link* to knative, I am not sure if we would go with `knative/pkg` or something else. It made sense to migrate the `tektoncd/operator` upstream to `knative/pkg` to be consistent with the rest of the projects — that's the main reason on why we did this. There is definitely some benefit to go the `knative/pkg` route (webhook self registring, easy cert setup, …), and there is definite drawbacks too (architecture knowledge, …).

Reply:
> There is definitely some benefit to go the knative/pkg route (webhook self registring, easy cert setup, …), 

When it comes to webhooks, it seems that operator-sdk is able to provide these capabilities via kubebuilder and https://github.com/jetstack/cert-manager. Other cert-manager solutions can be dropped in, assuming they similarly inject certs via an annotation.